### PR TITLE
Make DMD work with a struct as the main integral type

### PR DESCRIPTION
**Pull request by yebblies**
@redstar

### CODE REVIEW

**Comment by ibuclaw** [comment]
Seems reasonable, though this is proof of concept by the looks.

File: src/ddmd/builtin.d [hunk 1]
```diff
@@ -151,7 +151,7 @@ extern (C++) Expression eval_popcnt(Loc loc, FuncDeclaration fd, Expressions* ar
     int cnt = 0;
     while (n)
     {
-        cnt += (n & 1);
+        cnt += (n & 1) ? 1 : 0;
```
Comment (WalterBright) [inline src/ddmd/builtin.d]: What's the point of this?

File: src/ddmd/cppmangle.d [hunk 2]
```diff
@@ -734,7 +734,7 @@ public:
         }
         if (t.isConst())
             buf.writeByte('K');
-        buf.printf("A%llu_", t.dim ? t.dim.toInteger() : 0);
+        buf.printf("A%llu_", t.dim ? t.dim.toInteger() : dinteger_t(0));
```
Comment (WalterBright) [inline src/ddmd/cppmangle.d]: I don't understand the point of these casts.

**Comment by WalterBright** [comment]
What is the purpose of this?

**Comment by redstar** [comment]
It is the base of my cent/ucent implementation. As I can't use a cent/ucent I need to use a struct for emulation. This PR introduces the struct only.

**Comment by ibuclaw** [comment]
This is also the case for `real_t` too - all frontend code interacting with floating point already looks like this.

File: src/ddmd/dcast.d [hunk 3]
```diff
@@ -465,13 +465,13 @@ extern (C++) MATCH implicitConvTo(Expression e, Type t)
                 {
                     if (e.type.isunsigned())
                     {
-                        const f = real_t(value);
+                        const f = cast(real_t)value;
```
Comment (ibuclaw) [inline src/ddmd/dcast.d]: Potentially casting between two software emulation types, yummy. 😋

**Comment by yebblies** [comment]
Yes, the intent is to find all uses of implicit construction and conversion where a struct implementing 128bit support can't be dropped in. If we are going to move forward with cent/ucent then this will need to be done, and it will be easier to debug and review without the actual 128bit changes.

**Comment by andralex** [comment]
I like this, it's akin to dogofooding. Less magic keeps us all more honest. Any chance this can be revived?

**Comment by AndrewEdwards** [comment]
@ibuclaw @yebblies 

Besides being placed on the back-burner and completely forgotten, it is unclear why this PR is still open. @yebblies, please rebase. @ibuclaw, request finalization of review.

**Comment by ibuclaw** [comment]
The PR needs to be broken down into manageable pieces.  As and if I get round to it.  It's important for supporting `cent` types, and for using GCC types software integer types in GDC.

**Comment by ibuclaw** [comment]
Made some headway a number of years ago, rebased the branch on master now.

https://github.com/ibuclaw/dmd/tree/ctinteger

**Comment by dkorpel** [comment]
Closing this considering it is orphaned, and `cent`/`ucent` are deprecated now.

**Comment by deadalnix** [comment]
Reopening this.

Friend don't let friend make completely r*****ed decision such as deprecating `cent`/`ucent`. You might as well deprecate D with such a decision, unless you expect people to stop using 64bits CPU any time soon.

**Comment by dkorpel** [comment]
I understand you're upset that there's no plan to implement `cent` and `ucent`, but I don't see how re-opening an abandoned PR accomplishes anything. Perhaps you can make a case for it in the news group, or e-mail Mike Parker and ask if you can discuss this in a monthly Foundation meeting.

**Comment by ibuclaw** [comment]
This is work that we're going to do at some point in the future anyway, irrespective of `cent`/`ucent` actually being in the language, having internal N-bit emulation of integers at compile-time would make a few things easier.

**Pull request closed without merging**